### DISCUSSION TITLE
How to achieve concurrency for same workflow opened from different browser?

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#4961] [Design & tech]
Consider we have workflow which consist of 4 activities. User started workflow from one browser and perform some activity, lets say on activity 1 is completed and activity2 is Suspended for some reason. Now User logged in  from another browser and continued suspended activity 2 and then completed activity3. After this User once again came back to browser1, where user should get navigated to latest activity i.e. activity 4. So in this case how to achieve concurrency for same workflow opened from different browser?

**Comment by glime-ai** [comment]
To handle concurrency for the same workflow instance accessed from different browsers, you would need to ensure that the workflow state is consistently updated and that each browser session retrieves the latest state before attempting to resume or interact with the workflow. Elsa Workflow supports long-running workflows, which are designed to be suspended and resumed, potentially from different contexts such as different browser sessions.

Here are the steps you can take to achieve concurrency:

1. **Persist Workflow State**: Ensure that the workflow state is persisted after each activity execution. Elsa Workflow supports various persistence providers to store the workflow state.

2. **Retrieve Latest State**: Before resuming a workflow from a suspended state, retrieve the latest workflow state from the persistence store. This ensures that you are working with the most recent data.

3. **Handle Workflow Resumption**: When a user attempts to resume a workflow from a suspended activity, use the workflow instance ID to fetch the current state and determine the next activity to execute. If the workflow has moved past the suspended activity in another session, the user should be directed to the current activity.

4. **Use Distributed Locking**: If necessary, implement distributed locking to prevent race conditions and ensure that only one session can modify the workflow state at a time. This is particularly important if the workflow can be executed in parallel or if multiple users can interact with the same workflow instance concurrently.

5. **Signal Handling**: If the workflow uses signals to resume from suspension (e.g., waiting for an external event), ensure that signals are delivered to the correct workflow instance and that the instance is properly resumed.

6. **Client-Side Handling**: On the client side (in the browser), you may need to implement logic to periodically check for updates to the workflow state or to listen for server-sent events that indicate a state change.

7. **Workflow Design**: Design your workflow in such a way that it can handle being accessed from multiple sessions. This may involve using activities that are idempotent or that can handle being executed multiple times without adverse effects.

8. **Error Handling**: Implement error handling to manage scenarios where a user attempts to perform an action on an outdated workflow state.

Elsa Workflow's documentation and community discussions provide insights into handling concurrency, distributed locking, and workflow resumption. You may also consider looking into the Elsa Dashboard, which provides a user interface for managing and visualizing workflows, and could potentially be used to show users the current state of their workflows.

**Reply by NiveditaGalwade** [reply]
How to save workflow state after each activity execution when workflow created in programtic way. As of now I am saving workflow state in mongo db. 
how to retrieve latest state resuming a workflow from a suspended state?

**Reply by NiveditaGalwade** [reply]
Can someone please answer this question?

**Reply by douglasg14b** [reply]
I also would love an answer to this question... 💀

**Reply by NiveditaGalwade** [reply]
Can someone please address this question

**Reply by OvyDev** [reply]
In Elsa 2, you can call PersistWorkflow after the activity to set persistence behavior. However, by default the workflow is persisted after each burst (until it is blocking or is finished). So in your case, the worfklow is persisted when your activity return suspend.

**Comment by OvyDev** [comment]
I don't really understand the concurrency part of the question.

But to achieve what you want, I think you need a way to tell the next user about this workflow and how can their resume it. For example perhaps after activity 1, you can send email to 2nd user (or you can do this in a fork) and then activity 2. Or maybe instead of just sending email, you can add a record in a "Job" table in your db that can be queried by other users.
